### ISSUE TITLE
Gloo scatter gives wrong result for stride != 1

### ISSUE THREAD

**Issue by gaohuazuo** [pytorch/pytorch#24836]
## 🐛 Bug

`torch.distributed.scatter` on Gloo backend gives wrong result if input tensor stride != 1.

## To Reproduce

On worker 0:

```python
import torch.distributed
torch.distributed.init_process_group('gloo', world_size=2, rank=0, init_method='file:///tmp/pt-dist')
out = torch.zeros([3])
torch.distributed.scatter(out, [torch.linspace(0, 5, 6)[::2]]*2, 0)
print(out) # tensor([0., 1., 2.])
```

On worker 1:

```python
import torch.distributed
torch.distributed.init_process_group('gloo', world_size=2, rank=1, init_method='file:///tmp/pt-dist')
out = torch.zeros([3])
torch.distributed.scatter(out, [], 0)
print(out) # tensor([0., 1., 2.])
```

## Expected behavior

```
tensor([0., 2., 4.])
```

## Environment

PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] torch==1.2.0
[pip3] torchvision==0.4.0
[conda] Could not collect

**Comment by mrshenli** [comment]
@gaohuazuo good catch! Thanks for reporting, it seems missing a `contiguous()` call.

**Comment by mrshenli** [comment]
@gaohuazuo I took a look at `ProcessGroupGloo.cpp`, and it seems to me the the current implementation does not support non-contiguous input tensors. It requires the application to keep input tensors alive until `scatter` finishes. Even if it can create a new contiguous one within the `scatter(...)` method, it will be different from the inputs, and those async works only hold references which will not keep those tensors alive.  

Will it be acceptable if we just add a check to error out non-contiguous inputs?

**Comment by gaohuazuo** [comment]
I think checking `is_contiguous()` is actually better than calling `contiguous()` under the hood since this is a low level API.
